### ISSUE TITLE
Appending to sequence does not cause error

### ISSUE THREAD

**Issue by smackesey** [microsoft/pyright#2946]
From the [docs](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md)

>Switching from a mutable container type to a corresponding immutable container type is often an effective way to resolve type >errors relating to assignability. Let’s modify the example above by changing the type annotation for my_list_2.

>my_list_1: List[int] = [1, 2, 3]
>my_list_2: Sequence[Optional[int]] = my_list_1  # No longer an error
>my_list_2.append(None)  # Error

And indeed pyright reports an error as expected. However, this doesn't raise an error:

```
my_list_1: Sequence[int] = [1, 2, 3]
my_list_1.append(4)
```

Isn't this a false negative?

Running command line in language server mode, 1.1.215

**Comment by erictraut** [comment]
No, this isn't an error. It's intended behavior. The variable `my_list_1` is narrowed upon assignment to `list[int]`, and `append` is legal for a `list`.

**Comment by smackesey** [comment]
Seems to violate the principle of least surprise. If a user explicitly declares they want an immutable list (`Sequence` is declared as an immutable counterpart to `List` in your docs), what is the logic behind immediately silently replacing it with a mutable one? Aren't immutable lists a subset of lists? Then how can an immutable list be "narrowed" to a standard list?

Also FWIW `mypy` flags this as an error.

**Comment by erictraut** [comment]
Mypy contains inconsistent behavior when it comes to type narrowing on assignment. It doesn't do it when the assignment is on the same line as the type annotation. This has been discussed at length in the typing-sig, and there's pretty broad agreement that its behavior is inconsistent and wrong. There is a bug filed about it, but it hasn't been addressed in mypy yet.

You'll note that if you assign the list on a separate line, mypy no longer flags this as an error (which is correct).

```python
from typing import Sequence


my_list_1: Sequence[int]
my_list_1 = [1, 2, 3]
my_list_1.append(4)
```

Pyright isn't "silently replacing" the type. It's performing type narrowing, which is standard behavior and is type safe. The type `list[int]` is a subtype of `Sequence[int]` — i.e. the former is a narrower type of the latter. The concept of "mutable" and "immutable" are not part of the type system, and type checkers don't know anything about these concepts. Type checkers know about type relationships and type compatibility.

Consider the following:

```python
class A: ...
class B(A):
    def do_something(self) -> None: ...

def func1(val: A):
    if isinstance(val, B):
        reveal_type(val) # B
        val.do_something()

def func2():
    val: A = B()
    val.do_something()
```

**Comment by smackesey** [comment]
Thanks for your explanation. I understand how the narrowing machine is operating here-- yes, lists are a subtype of `Sequence` in Python, and so following this logic the narrowing makes sense.

However, this does not make sense to me:

>The concept of "mutable" and "immutable" are not part of the type system, and type checkers don't know anything about these concepts. 

This statement is confusing in light of the docs I linked, which has a whole section describing "mutable" and "immutable" container types. The docs literally illustrate how in one case using the mutable type causes Pyright to raise a type error, and the immutable type doesn't. How could this be true if "type checkers don't know anything about these concepts"?

Basically there is an ambiguity in the use of `Sequence` between this discussion and your docs. Your docs specifically contrast `List` and `Sequence` as mutable/immutable counterparts. Using these definitions, "narrowing" from Sequence to List makes no sense, for the simple reason that mutable lists are not a subtype of immutable lists.

But I understand that because the Python interpreter does not enforce immutability of Sequences, it's type hierarchy does not account for this, and so the narrowing makes sense through that lens.

In any case I think the linked docs should be clarified.
